src/main/electron-types.ts

```diff
@@ -2,23 +2,84 @@ import * as path from 'node:path';
 
 import { ElectronVersions } from '@electron/fiddle-core';
 import fetch from 'cross-fetch';
-import { IpcMainEvent, app } from 'electron';
+import { BrowserWindow, IpcMainEvent, app } from 'electron';
 import * as fs from 'fs-extra';
+import watch from 'node-watch';
 import packageJson from 'package-json';
 import readdir from 'recursive-readdir';
 import semver from 'semver';
 
 import { ipcMainManager } from './ipc';
-import { NodeTypes } from '../interfaces';
+import { ELECTRON_DTS } from '../constants';
+import { NodeTypes, RunnableVersion, VersionSource } from '../interfaces';
 import { IpcEvents } from '../ipc-events';
 
 let electronTypes: ElectronTypes;
 
 export class ElectronTypes {
+  private localPaths: Map<BrowserWindow, string>;
+  private watchers: Map<string, fs.FSWatcher>;
+
   constructor(
     private readonly knownVersions: ElectronVersions,
+    private readonly electronCacheDir: string,
     private readonly nodeCacheDir: string,
-  ) {}
+  ) {
+    this.localPaths = new Map();
+    this.watchers = new Map();
+  }
+
+  private getWindowsForLocalPath(localPath: string): BrowserWindow[] {
+    return Array.from(this.localPaths.entries())
+      .filter(([, path]) => path === localPath)
+      .map(([window]) => window);
+  }
+
+  public async getElectronTypes(
+    window: BrowserWindow,
+    ver: RunnableVersion,
+  ): Promise<string | undefined> {
+    const { localPath: dir, source, version } = ver;
+    let content: string | undefined;
+
+    // If it's a local development version, pull Electron types from out directory.
+    if (dir) {
+      const file = path.join(dir, 'gen/electron/tsc/typings', ELECTRON_DTS);
+      content = this.getTypesFromFile(file);
+      try {
+        this.unwatch(window);
+        this.localPaths.set(window, dir);
+
+        // If no watcher for that path yet, create it
+        if (!this.watchers.has(dir)) {
+          const watcher = watch(file, () => {
+            // Watcher should notify all windows watching that path
+            const windows = this.getWindowsForLocalPath(dir);
+            for (const window of windows) {
+              ipcMainManager.send(
+                IpcEvents.ELECTRON_TYPES_CHANGED,
+                [fs.readFileSync(file, 'utf8'), version],
```

Comment:
I think we already have the file contents from L48, so it could be reused here instead of re-reading the file

Reply:
This line is inside the callback for the file watcher, so the file contents have changed and need to be read again. I can pull this out into a private function to make that more clear.

Reply:
Ah, makes sense. I think it's fine as it is then. One question though, won't this throw if the file is deleted or does the callback only run when the file exists?

Reply:
nevermind, I don't think it can be deleted 😅

Reply:
It's a good point, though. Prior to refactoring the read was in a `try/catch` block, now it's not - I'll fix that.

https://github.com/electron/fiddle/blob/47f62e36767f43846645b7bee9eab7e70d0aa55e/src/renderer/electron-types.ts#L89-L97

Reply:
Refactored in 636e784831e5c5fc20d6e3cb5b82d6dff7a92173 - also noticed that the file contents were being read in the for-loop unnecessarily so cleaned that up as well. Thanks for the comment! 🙂